NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp

```diff
@@ -553,319 +536,337 @@ void CCpuMathEngine::BlobConvolution( const CConvolutionDesc& convDesc, const CC
-
-	int filterChannels = filter.Depth() * filter.Channels();
-
-	int outputLineStart;
-	int outputLineEnd;
-	OmpGetTaskIndexAndCount( output.ObjectCount() * output.Height(), outputLineStart, outputLineEnd );
-	outputLineEnd += outputLineStart;
-	for( int step = outputLineStart; step < outputLineEnd; ++step ) {
-		float* outputDataPtr = outputDataRaw + step * output.Width() * output.Depth() * output.Channels();
-
-		if( freeTermData != nullptr ) {
-			// Set the free term
-			setVectorToMatrixRows( outputDataPtr, output.Width(), output.Depth() * output.Channels(), freeTermDataRaw );
-		} else {
-			vectorFill( outputDataPtr, 0, output.Width() * output.Depth() * output.Channels() );
-		}
+//------------------------------------------------------------------------------------------------------------
 
-		int batch = step / output.Height();
-		int row = step % output.Height();
-		int inputRowStart = (row + desc.PaddingHeight - filter.Height() + desc.StrideHeight) / desc.StrideHeight;
-		if(inputRowStart < 0) {
-			inputRowStart = 0;
-		}
-		int filterRowBackStart = row - inputRowStart * desc.StrideHeight + desc.PaddingHeight;
-		if(0 > filterRowBackStart || filterRowBackStart >= filter.Height()) {
-			continue;
-		}
-		int filterRowBackEnd = filter.Height() + row - output.Height() - desc.PaddingHeight;
-		if(filterRowBackEnd < 0) {
-			filterRowBackEnd = 0;
-		}
+class CRoundRobinBuffer final {
+public:
+	explicit CRoundRobinBuffer( IMathEngine& mathEngine, int numItems, int itemSize ) :
+		temp( mathEngine, /*bytes-size*/numItems * itemSize ), raw( NeoML::GetRaw( temp.GetHandle() ) ),
+		numItems( numItems ), itemSize( itemSize ), currPos( 0 ),
+		prevBatchedSourceRowStart( -1 ), prevNumItemsToAdd( 0 ), startItem( 0 )
+	{}
 
-		int inputRow = inputRowStart;
-		for(int filterRow = filterRowBackStart;
-			filterRow >= filterRowBackEnd;
-			filterRow -= desc.StrideHeight, ++inputRow) {
-			// The temp blob stores the filter rows multiplied by input; add them to the output rows in correct positions
-			const float* tempRowData = tempRaw + ( ( batch * input.Height() + inputRow )
-				* input.Width() * filter.Height() + filterRow ) * filter.Width() * filterChannels;
-
-			for(int col = -desc.PaddingWidth;
-				col <= output.Width() + desc.PaddingWidth - filter.Width();
-				col += desc.StrideWidth) {
-
-				int tempRowDataShift = 0;
-				int toCopy = filter.Width();
-				int pos = col;
-				if(pos < 0) {
-					tempRowDataShift = -pos;
-					toCopy += pos;
-					pos = 0;
-				}
-				if(pos + toCopy > output.Width()) {
-					toCopy = output.Width() - pos;
-				}
-				if( toCopy > 0 ) {
-					toCopy *= filterChannels;
-					tempRowDataShift *= filterChannels;
-					float* outputVec = outputDataPtr + pos * filterChannels;
-					vectorAdd( outputVec, tempRowData + tempRowDataShift, outputVec, toCopy );
-				}
-				tempRowData += filter.Height() * filter.Width() * filterChannels;
-			}
-		}
+	int GetExistItemsNum( int batchedSourceRowStart, int numItemsToAdd )
```

Comment:
This method `Get`s something but `numItemsToAdd` says that something will be added... This class need _a lot_ of comments.

Reply:
fixed